### DISCUSSION TITLE
Can you define methods to call during runtime of a module?

### DISCUSSION THREAD

**Discussion by Alexzander-Stone** [aptos-labs/aptos-developer-discussions#159] [Questions] [answered]
### Describe your question in detail.

Asked by `Lokesh | LYNC` on Discord:
>hey guys, I have a question: In a move script, can we pass in like a module name, function name, and args. Something like this:
> ```
> script {
>    use std::signer;
>    use std::string::{Self, String};
>
>    fun main(
>        src: &signer,
>        module_addr: address,
>        module_name: String,
>        func_name: String,
>    ) {
>        @module_addr::module_name::func_name();
>
>    }
> }
> ```
>Is this possible or not? 

### What have you tried or looked at?  Or how can we reproduce the error?

Code given.

### Which operating system are you using?

N/A

### Which SDK or tool are you using? (if any)

N/A

**Comment by Alexzander-Stone** [comment, marked as answer]
**No, this is not possible. Aptos Move uses [static dispatch](https://aptos.dev/concepts/move-on-aptos/#comparison-to-other-vms:~:text=Function%20calling-,Static%20dispatch,-Static%20dispatch), as of April 23rd, 2024.**

------

You'll receive errors from the Aptos CLI when attempting to compile the code, like:
```
error[E03001]: address with no value
   ┌─ /Users/user/Documents/aptos/aptos-dev-discussions-playground/can_you_dynamic_dispatch/sources/dispatch_test.move:11:8
   │
11 │        module_addr::module_name::func_name();
   │        ^^^^^^^^^^^ address 'module_addr' is not assigned a value

error[E03002]: unbound module
   ┌─ /Users/user/Documents/aptos/aptos-dev-discussions-playground/can_you_dynamic_dispatch/sources/dispatch_test.move:11:8
   │
11 │        module_addr::module_name::func_name();
   │        ^^^^^^^^^^^^^^^^^^^^^^^^ Unbound module 'module_addr::module_name'
```
Notice that the error originated from `dispatch_test.move`, hinting at the static dispatch limitation.

### Static vs Dynamic Dispatch
[Static dispatch](https://en.wikipedia.org/wiki/Static_dispatch) determines which operations, or logic implementations, to perform during the compilation step.
[Dynamic dispatch](https://en.wikipedia.org/wiki/Dynamic_dispatch) allows for operations to be determined during runtime. 

### Applied to your example
The values of `module_addr`, `module_name`, and `func_name` are unknown until the method is actively ran.
During compilation, the process wouldn't know which function's logic to implement since each value is unknown, leading to failure.

> This is why it would work to explicitly define it within our code, since our compilation step would be given exact values, rather than unknown values.

To support your given concept, the dispatch steps would need to be performed during run time, when the values of the above variables are available. 

### You might ask, "How do DEXs/{other tech} work, then?"
While Aptos Move doesn't have the ability to determine the implementation of a function during runtime, it doesn't mean that we can't pass dynamic data.

For example, type arguments are commonly used to describe the type of the Coins being swapped on DEXs. This is valid since the function implementation isn't changing between different types, but rather the instance of the pre-defined type.

**Reply by gregnazario** [reply]
You can also use dynamic dispatch with dispatchable fungible assets:

https://aptos.dev/en/build/smart-contracts/fungible-asset#dispatchable-fungible-asset-advanced
